### PR TITLE
JSDoc: String Module

### PR DESCRIPTION
**Pull request by poteat**
Provide JSDoc for all string module utilities.

### CODE REVIEW

File: style-guide/jsdoc.md [hunk 1]
```diff
@@ -13,6 +13,8 @@ export interface Utility extends Kind.Kind {
 In this case, we document both the type and the interface. Usually, the contents of the JSDoc are similar between them.
 
 ```ts
+import { $, Kind } from 'ts-toolbelt'
```
Comment (MajorLift) [inline style-guide/jsdoc.md:16]: This seems like a typo.😁

File: style-guide/jsdoc.md [hunk 2]
```diff
@@ -38,37 +40,83 @@ export interface Length extends Kind.Kind {
...
+
+In this example, `EndsWith_T` is the intermediate interface for currying and is not exported, so it doesn't have a JSDoc comment. The examples for `_$endsWith` and `EndsWith` demonstrate the correct usage of 2-ary kinds, providing all parameters.
 
 ## Case of multi-stage kinds
 
-Some kinds are quite complex, and are defined via long sequences of generic parameters. In these cases, each stage should be documented. The following is  a simplified example.
+Some kinds are quite complex, and are defined via long sequences of generic parameters. In these cases, each template should be documented. The following is a simplified example.
 
 ```ts
+import { $, Kind } from 'ts-toolbelt'
```
Comment (MajorLift) [inline style-guide/jsdoc.md:104]: This 3

File: src/string/to-upper.ts [hunk 3]
```diff
@@ -1,7 +1,23 @@
...
+ *
+ * @example
+ * type T0 = String._$toUpper<'foo'> // 'FOO'
+ * type T1 = String._$toUpper<'bar'> // 'BAR'
+ */
 export type _$toUpper<S extends string> = Uppercase<S>
 
+/**
+ * `String.ToUpper` is a type-level function that converts a string to uppercase.
+ *
```
Comment (MajorLift) [inline src/string/to-upper.ts:16]: `@template` seems to be missing
```suggestion
 * @template S - The string to convert to uppercase.
```

File: src/string/to-lower.ts [hunk 4]
```diff
@@ -1,7 +1,23 @@
...
+ *
+ * @example
+ * type T0 = String._$toLower<'HELLO'> // 'hello'
+ * type T1 = String._$toLower<'WORLD'> // 'world'
+ */
 export type _$toLower<S extends string> = Lowercase<S>
 
+/**
+ * `String.ToLower` is a type-level function that converts a string to lowercase.
+ *
```
Comment (MajorLift) [inline src/string/to-lower.ts:16]: `@template` tag missing

File: src/string/tail.ts [hunk 5]
```diff
@@ -8,6 +17,13 @@ export type _$tail<S extends string> = S extends `${string}${infer Tail}`
   ? S
   : ''
 
+/**
+ * `String.Tail` is a type-level function that extracts every element after the first element of a string.
+ *
```
Comment (MajorLift) [inline src/string/tail.ts:22]: missing `@template` tag

File: src/string/starts-with.ts [hunk 6]
```diff
@@ -9,6 +19,13 @@ interface StartsWith_T<Prefix extends string> extends Kind.Kind {
   f(x: Type._$cast<this[Kind._], string>): _$startsWith<Prefix, typeof x>
 }
 
+/**
+ * `String.StartsWith` is a type-level function that checks if a string starts with a given prefix.
+ *
```
Comment (MajorLift) [inline src/string/starts-with.ts:24]: missing `@template` tags

File: style-guide/jsdoc.md (hunk 1 above)
Comment (poteat) [reply style-guide/jsdoc.md:16]: Ah definitely, good catch!

File: src/string/slice.ts [hunk 7]
```diff
@@ -9,6 +19,13 @@ interface Slice_T<N extends Number.Number> extends Kind.Kind {
   f(x: Type._$cast<this[Kind._], string>): _$slice<typeof x, N>
 }
 
+/**
+ * `String.Slice` is a type-level function that slices a string from a given index.
+ *
```
Comment (MajorLift) [inline src/string/slice.ts:24]: missing `@template` tag

File: src/string/slice.ts [hunk 8]
```diff
@@ -1,5 +1,15 @@
-import { Kind, Type, String, Number, List } from '..'
+import { List, Kind, Type, Number, String } from '..'
```
Comment (MajorLift) [inline src/string/slice.ts:2]: I think we might want to make a breaking change so this corresponds to `List.Slice`. What do you think?

File: src/string/replace.ts [hunk 9]
```diff
@@ -21,14 +35,30 @@ export type _$replace<
...
+/**
+ * `Replace_T` is an intermediate interface for currying.
+ */
 interface Replace_T<From extends string> extends Kind.Kind {
   f(x: Type._$cast<this[Kind._], string>): Replace_T2<From, typeof x>
 }
 
+/**
+ * `String.Replace` is a type-level function that replaces all instances of a string with another string.
+ *
```
Comment (MajorLift) [inline src/string/replace.ts:54]:
```suggestion
 *
 * @template S - The string to replace in.
```

File: src/string/replace.ts [hunk 10]
```diff
@@ -21,14 +35,30 @@ export type _$replace<
...
+ * @template From - The string to replace.
+ * @template To - The string to replace with.
+ *
```
Comment (MajorLift) [inline src/string/replace.ts:57]:
```suggestion
 * @template S - The string to replace in.
 *
```

File: src/string/first.ts [hunk 11]
```diff
@@ -1,11 +1,27 @@
...
+ */
 export type _$first<S extends string> = S extends `${infer Head}${string}`
   ? Head
   : string extends S
   ? S
   : ''
 
+/**
+ * `String.First` is a type-level function that extracts the first character from a string.
+ *
```
Comment (MajorLift) [inline src/string/first.ts:20]:
```suggestion
 * @template S - The string to extract the first character from.
 *
```

File: src/string/ends-with.ts [hunk 12]
```diff
@@ -9,6 +19,13 @@ interface EndsWith_T<T extends string> extends Kind.Kind {
   f(x: Type._$cast<this[Kind._], string>): _$endsWith<T, typeof x>
 }
 
+/**
+ * `String.EndsWith` is a type-level function that checks if a string ends with a given suffix.
+ *
```
Comment (MajorLift) [inline src/string/ends-with.ts:24]:
```suggestion
 *
 * @template Suffix - The suffix to check for.
 * @template S - The string to check.
 *
```

File: src/string/append.ts [hunk 13]
```diff
@@ -1,11 +1,30 @@
...
 
 interface Append_T<Suffix extends string> extends Kind.Kind {
   f(x: Type._$cast<this[Kind._], string>): _$append<Suffix, typeof x>
 }
 
+/**
+ * `String.Append` is a type-level function that appends a suffix to a string.
+ *
+ * @template Suffix - The string to append.
+ *
```
Comment (MajorLift) [inline src/string/append.ts:23]:
```suggestion
 * @template S - The original string.
 *
```

**Review by MajorLift — APPROVED** [review]
LGTM!

File: src/string/slice.ts [hunk 14]
```diff
@@ -9,6 +19,15 @@ interface Slice_T<N extends Number.Number> extends Kind.Kind {
   f(x: Type._$cast<this[Kind._], string>): _$slice<typeof x, N>
 }
 
+/**
+ * `String.Slice` is a type-level function that slices a string from a given index.
```
Comment (MajorLift) [inline src/string/slice.ts:23]:
```suggestion
 * `String.Slice` is a type-level function that slices a string from a given index.
 *
```

File: src/string/starts-with.ts [hunk 15]
```diff
@@ -9,6 +19,15 @@ interface StartsWith_T<Prefix extends string> extends Kind.Kind {
   f(x: Type._$cast<this[Kind._], string>): _$startsWith<Prefix, typeof x>
 }
 
+/**
+ * `String.StartsWith` is a type-level function that checks if a string starts with a given prefix.
```
Comment (MajorLift) [inline src/string/starts-with.ts:23]:
```suggestion
 * `String.StartsWith` is a type-level function that checks if a string starts with a given prefix.
 *
```

File: src/string/tail.ts [hunk 16]
```diff
@@ -8,6 +17,14 @@ export type _$tail<S extends string> = S extends `${string}${infer Tail}`
   ? S
   : ''
 
+/**
+ * `String.Tail` is a type-level function that extracts every element after the first element of a string.
```
Comment (MajorLift) [inline src/string/tail.ts:21]:
```suggestion
 * `String.Tail` is a type-level function that extracts every element after the first element of a string.
 *
```

File: src/string/to-lower.ts [hunk 17]
```diff
@@ -1,7 +1,24 @@
...
+ * @template S - The string to convert to lowercase.
+ *
+ * @example
+ * type T0 = String._$toLower<'HELLO'> // 'hello'
+ * type T1 = String._$toLower<'WORLD'> // 'world'
+ */
 export type _$toLower<S extends string> = Lowercase<S>
 
+/**
+ * `String.ToLower` is a type-level function that converts a string to lowercase.
```
Comment (MajorLift) [inline src/string/to-lower.ts:15]:
```suggestion
 * `String.ToLower` is a type-level function that converts a string to lowercase.
 *
```

**Pull request merged**